packages/datadog-plugin-aws-sdk/src/index.js

```diff
@@ -3,67 +3,100 @@
 
-function createWrapRequest (tracer, config) {
-  config = normalizeConfig(config)
-  return function wrapRequest (send) {
-    return function requestWithTrace (cb) {
-      if (!this.service) return send.apply(this, arguments)
+const requestTags = new WeakMap()
 
-      const serviceIdentifier = this.service.serviceIdentifier
+class AwsSdkPlugin extends Plugin {
+  static get name () {
+    return 'aws-sdk'
+  }
 
-      if (!awsHelpers.isEnabled(serviceIdentifier, config[serviceIdentifier], this)) {
-        return send.apply(this, arguments)
-      }
+  constructor (...args) {
+    super(...args)
 
-      const serviceName = getServiceName(serviceIdentifier, tracer, config)
-      const childOf = tracer.scope().active()
+    this.addSub('apm:aws:request:start', ({ request, serviceIdentifier, operation, awsRegion, awsService }) => {
+      if (!awsHelpers.isEnabled(serviceIdentifier, this.config[serviceIdentifier], request)) {
+        return
+      }
+      const serviceName = getServiceName(serviceIdentifier, this.tracer, this.config)
+      const childOf = this.tracer.scope().active()
       const tags = {
         [Tags.SPAN_KIND]: 'client',
         'service.name': serviceName,
-        'aws.operation': this.operation,
-        'aws.region': this.service.config && this.service.config.region,
-        'aws.service': this.service.api && this.service.api.className,
+        'aws.operation': operation,
+        'aws.region': awsRegion,
+        'aws.service': awsService,
         'component': 'aws-sdk'
       }
+      requestTags.set(request, tags)
 
-      const span = tracer.startSpan('aws.request', {
-        childOf,
-        tags
-      })
+      const span = this.tracer.startSpan('aws.request', { childOf, tags })
 
-      this.on('complete', response => {
-        if (!span) return
+      analyticsSampler.sample(span, this.config.measured)
 
-        awsHelpers.addResponseTags(span, response, serviceIdentifier, config, tracer)
-        awsHelpers.finish(config, span, response, response.error)
-      })
+      awsHelpers.requestInject(span, request, serviceIdentifier, this.tracer)
 
-      analyticsSampler.sample(span, config.measured)
+      const store = storage.getStore()
 
-      awsHelpers.requestInject(span, this, serviceIdentifier, tracer)
+      this.enter(span, store)
+    })
 
-      const request = this
+    this.addSub('apm:aws:request:complete', ({ response, serviceIdentifier }) => {
+      const store = storage.getStore()
+      if (!store) return
+      const { span } = store
+      if (!span) return
+      awsHelpers.addResponseTags(span, response, serviceIdentifier)
+      awsHelpers.finish(this.config, span, response, response.error)
+    })
 
-      return tracer.scope().activate(span, () => {
-        if (typeof cb === 'function') {
-          arguments[0] = awsHelpers.wrapCb(cb, serviceIdentifier, tags, request, tracer, childOf)
+    this.addSub('apm:aws:response', obj => {
```

Comment:
Handling this for each service that needs it instead of in a generic way would be more flexible and would avoid most of this code. There could even be separate plugins for each AWS service, which would make it easier to enable/disable them individually which is supported for message queues.